semantics/language/execution/io-buffered.k

```diff
@@ -102,217 +88,188 @@ module C-IO-BUFFERED
      rule locations(.List) => .Set
-     rule locations(ListItem(bwrite(Loc:SymLoc, _)) L:List)
+     rule locations(ListItem(bwrite(Loc::SymLoc, _)) L::List)
           => SetItem(Loc) locations(L)
 
-     rule <k> readBytes(Loc:SymLoc, Size:Int, T:Type)
+     rule <k> readBytes(Loc::SymLoc, Size::Int, T::Type)
                => checkEffectiveType(T, getEffectiveType(Loc))
                ~> readBytes-aux(Loc, Size, .List, T)
           ...</k>
-          <thread-id> ThreadId:Int </thread-id>
+          <thread-id> ThreadId::Int </thread-id>
           requires notBool ((isThreadDuration(Loc) orBool isAutoDuration(Loc))
                     andBool (ThreadId =/=K getThreadId(Loc)))
           [structural]
      rule <k> (.K => UNSPEC("EIO6", "Attempting to access an object with thread or auto storage duration in a thread other than the one in which the object is associated.", "6.2.4:4-5"))
-               ~> readBytes(Loc:SymLoc, _, _)
+               ~> readBytes(Loc::SymLoc, _, _)
           ...</k>
-          <thread-id> ThreadId:Int </thread-id>
+          <thread-id> ThreadId::Int </thread-id>
           requires (isThreadDuration(Loc) orBool isAutoDuration(Loc))
                andBool (ThreadId =/=K getThreadId(Loc))
           [structural]
 
      // loc, size in bytes, aux list
      syntax KItem ::= "readBytes-aux" "(" SymLoc "," Int "," List "," Type ")"
-     rule readBytes-aux(Loc:SymLoc, Size:Int, Aux:List, T:Type)
-          => readByte(Loc, T)
-          ~> readBytes-aux(Loc +bytes 1, Size -Int 1, Aux, T)
-          requires Size >Int 0
-          [structural]
-     rule readBytes-aux(_, 0, Aux:List, _) => dataList(bvalues(Aux))
+     rule readBytes-aux(_, 0, Aux::List, _) => dataList(Aux)
           [structural]
-     rule (tv(V:CValue, T:Type) => .K)
-          ~> readBytes-aux(_, _,
-               (Aux:List => Aux ListItem(tv(V, T))), _)
-          [structural]
-
-     syntax List ::= bvalues(List) [function]
-     rule bvalues(ListItem(tv(K:K, _)) L:List) => ListItem(K) bvalues(L)
-     rule bvalues(.List) => .List
 
      syntax Error ::= "errorReadOutOfBounds"
                     | "errorUnseqEffects"
                     | "errorNonVolatileLVal"
 
      syntax KItem ::= readByte(SymLoc, Type)
-     rule <k> readByte(loc(Base:SymBase, Offset:Int), T:Type)
+     rule <k> readBytes-aux(loc(Base::SymBase, Offset::Int), Size::Int, Aux::List, T::Type)
                // TODO(chathhorn): do we really need to check this stuff every
                // byte?
-               => assert(notBool (loc(Base, Offset) in Locs),
-                    errorUnseqEffects)
-               ~> assert(Offset <Int Len, errorReadOutOfBounds)
-               ~> assert((notBool (Volatile in getQualsAtOffset(ObjT, Offset)))
-                    orBool isVolatileType(T),
-                    errorNonVolatileLVal)
-               ~> tv(V, t(.Set, no-type))
+               => #if notBool (loc(Base, Offset) in Locs) #then .K #else
```

Comment:
At first sight, these rules look rather ugly and repetitive, both before and after @dwightguth 's changes.  Although the side conditions will be evaluated only after the matching takes place, and probably once that happens the side conditions hold with high probability, it still looks ugly and potentially inefficient to repeat the same side condition Size > 0 in each rule.  Maybe you guys can find a different way to write them all together, which would satisfy you both?

Orthogonally to that, the issue of writing rules like
```
1:
rule <k> readByte(stuff) => .K ...</k> <mem> sideEffects </mem> 
rule readBytes(N) => readByte(stuff) ~> readBytes(N - 1)
rule readBytes(0) => .K
```
versus
```
2:
rule <k> readBytes(N) => readBytes(N - 1) ...</k> <mem> sideEffects </mem>
rule readBytes(0) => .K
```
remains.

The benefit of 1 is that the rules look more "functional", or "inductive", but more rewrite steps are generated.
The benefit of 2 is performance of execution and fewer rules, but the rules are harder to reason with.

This is reminiscent of forwards recursive (1) versus tail recursive (2) functions in functional programming.  Although it is harder to reason about programs which are written in a tail recursive fashion, the performance benefits are often non-negligible and people prefer them over the ordinary recursive variants.  So in the end it boils down as a matter of subjective taste and preference.  Speaking of which, I would have probably preferred 2, too, because it gives me an opportunity to show off K's compact notation a bit more:
```
3:
rule <k> readBytes(N => N - 1) ...</k> <mem> sideEffects </mem>
rule readBytes(0) => .K
```

@andreistefanescu , how much does this affect you with your proofs?  If it is more a matter of taste and principle rather than technical, I would go with 2 or 3, because I think we will need to refactor the C semantics eventually anyway, and then we may rethink some of these issues, but at least, until then, we have a faster undefinedness checker.